### PR TITLE
Implemented the slider bar present on GHWT guitar controllers

### PR DESCRIPTION
**Pull request by Freyaday**
This change is [Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/4418)

### CODE REVIEW

**Comment by RisingFog** [comment]
@dolphin-emu-bot rebuild

**Comment by JosJuice** [comment]
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp, line 58 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418#-KVeUgFH2Qr7Mq2RCgzx:-KVeUgFH2Qr7Mq2RCgzy:b-cwssu1) ([raw file](https://github.com/dolphin-emu/dolphin/blob/98f49403d84e1b63ad5fad110498b8728e33a8e8/Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp#L58)):*
> ```cpp
> 
> 	//touchbar
> 	groups.emplace_back(m_touchbar = new Slider(_trans("TouchBar")));
> ```

Please call this by the official name "Slider Bar" instead of "TouchBar". (Renaming m_touchbar isn't as important, only the string that's exposed to users.)

---

*[Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp, line 59 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418#-KVeUlhWIM2xZgTcm5fz:-KVeUlhWIM2xZgTcm5g-:b7orp7g) ([raw file](https://github.com/dolphin-emu/dolphin/blob/98f49403d84e1b63ad5fad110498b8728e33a8e8/Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp#L59)):*
> ```cpp
> 	//touchbar
> 	groups.emplace_back(m_touchbar = new Slider(_trans("TouchBar")));
> 	//m_touchbar->controls.emplace_back(new ControlGroup::Input(_trans("Bar")));
> ```

Why is this commented out?

---

*[Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.h, line 42 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418#-KVeVFUU8Mo4JPjaQRus:-KVeVFUVo0em7AdW5a2n:b-gay3yn) ([raw file](https://github.com/dolphin-emu/dolphin/blob/98f49403d84e1b63ad5fad110498b8728e33a8e8/Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.h#L42)):*
> ```objc
> 	Slider* m_touchbar;
> 
> 	const std::map <const ControlState, const u8> m_controlCodes
> ```

Please put this `map` in the .cpp file. There's no need for other code to know these exact values.

Also, the name could use some improvement. It should preferably include slider_bar/touch_bar. (Note that our standard is to use underscores to separate words, not capitalization.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418)*

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by JosJuice** [comment]
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[.gitignore, line 37 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418#-KVeZd8KDcKI8lSZEM-G:-KVeZd8LPn4NxwL4zy1t:b-y37u55) ([raw file](https://github.com/dolphin-emu/dolphin/blob/72e91bf474a54401e1caf2ea30b71a96e782fbda/.gitignore#L37)):*
> ```
> # Ignore IDEA/Clion files/dirs
> /.idea/
> /Source/dolphin-emu.sln.db
> ```

Please group this with the other VS files (above the line "# Ignore build info file created by QtCreator").

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418)*

**Comment by JosJuice** [comment]
You will need to squash your commits into one instead of having additional fixup commits.

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by BhaaLseN** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp, line 12 at r4](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418#-KVfAorfIOhqZhjc6VqK:-KVfAorguYOSQp0ROKgV:bkphzji) ([raw file](https://github.com/dolphin-emu/dolphin/blob/74115837a80fd088e6861683f40666fd39132a7b/Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp#L12)):*
> ```cpp
> #include "Core/HW/WiimoteEmu/WiimoteEmu.h"
> 
> static const std::map<const ControlState, const u8> s_touchbar_control_codes{
> ```

This table looks...meh. I guesstimate (with some weak Excel skills) that the range `-1 .. 1` is mapped to `0x00 .. 0x1F` in a linear fashion and could be calculated directly from the Slider value rather than using this LUT. Do you think you could give that a try?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418)*

**Comment by Freyaday** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp, line 12 at r4](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418#-KVfAorfIOhqZhjc6VqK:-KVghqhPW003cI0v2HBk:b-f1bjle) ([raw file](https://github.com/dolphin-emu/dolphin/blob/74115837a80fd088e6861683f40666fd39132a7b/Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp#L12)):*
<details><summary><i>Previously, BhaaLseN (BhaaL) wrote…</i></summary>
> This table looks...meh. I guesstimate (with some weak Excel skills) that the range `-1 .. 1` is mapped to `0x00 .. 0x1F` in a linear fashion and could be calculated directly from the Slider value rather than using this LUT. Do you think you could give that a try?
</details>
I did exactly that at first. Turns out,it's not a linear mapping at all. It's actually cubic: https://puu.sh/s541j/c2a71e1281.png

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418)*

**Comment by JMC47** [comment]
Can you note what Guitar Hero games use this variant of the controller for Wiki + testing purposes.  I can handle updating everything, but, i'm not a huge guitar hero player so I don't know all the nuances like this.

**Comment by Freyaday** [comment]
The Slider Bar is new to the Guitar Hero World Tour guitars and was carried over into Guitar Hero 5. It was not present in GH3, but the game doesn't check that byte as far as I know and we're already telling the game that we have a GHWT guitar plugged in with nobody touching the slider bar so things should be good on that front.

---

Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418#-:-KVgkcBlnBkMma6j2Lba:bh936zn)*

**Comment by lioncash** [comment]
Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp, line 58 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418#-KVeUgFH2Qr7Mq2RCgzx:-KVgp_hyovM6h1yHavEr:bgz2ac7) ([raw file](https://github.com/dolphin-emu/dolphin/blob/98f49403d84e1b63ad5fad110498b8728e33a8e8/Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp#L58)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> Please call this by the official name "Slider Bar" instead of "TouchBar". (Renaming m_touchbar isn't as important, only the string that's exposed to users.)
</details>
The variable should likely be changed to `m_sliderbar` or `m_slider_bar` as well

---

*[Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp, line 11 at r4](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418#-KVgpsn80Wj_A4PUSsIl:-KVgpsn9f16wpOh0p62b:b-ug78tv) ([raw file](https://github.com/dolphin-emu/dolphin/blob/74115837a80fd088e6861683f40666fd39132a7b/Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp#L11)):*
> ```cpp
> #include "Core/HW/WiimoteEmu/Attachment/Guitar.h"
> #include "Core/HW/WiimoteEmu/WiimoteEmu.h"
> ```

Did you add trailing spaces throughout the whole file or something? Compared to the previous diff, it looks like you did, given these whitespace dots all over the place.

---

*[Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp, line 12 at r4](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418#-KVgpEaoNznT3h_A8_fe:-KVgpEapOCEUKMpriewG:b-ldh9fh) ([raw file](https://github.com/dolphin-emu/dolphin/blob/74115837a80fd088e6861683f40666fd39132a7b/Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp#L12)):*
> ```cpp
> #include "Core/HW/WiimoteEmu/WiimoteEmu.h"
> 
> static const std::map<const ControlState, const u8> s_touchbar_control_codes{
> ```

This map should be within the WiimoteEmu namespace below. You also need to include `<map>`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418)*

**Comment by Freyaday** [comment]
Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp, line 11 at r4](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418#-KVgpsn80Wj_A4PUSsIl:-KVhEU45NnIywGCtAHKy:b7nnnxk) ([raw file](https://github.com/dolphin-emu/dolphin/blob/74115837a80fd088e6861683f40666fd39132a7b/Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp#L11)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> Did you add trailing spaces throughout the whole file or something? Compared to the previous diff, it looks like you did, given these whitespace dots all over the place.
</details>
Oh, I bet somewhere along the line something decided to change all the line endings to CFLR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418)*

**Comment by Freyaday** [comment]
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp, line 12 at r4](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418#-KVgpEaoNznT3h_A8_fe:-KVhH6GHZOb9OEekMBtX:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/74115837a80fd088e6861683f40666fd39132a7b/Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp#L12)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> This map should be within the WiimoteEmu namespace below. You also need to include `<map>`.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418)*

**Comment by BhaaLseN** [comment]
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp, line 12 at r4](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418#-KVfAorfIOhqZhjc6VqK:-KVjbUb9nK-FzrFqkR0-:b-l2pknl) ([raw file](https://github.com/dolphin-emu/dolphin/blob/74115837a80fd088e6861683f40666fd39132a7b/Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp#L12)):*
<details><summary><i>Previously, Freyaday wrote…</i></summary>
> I did exactly that at first. Turns out,it's not a linear mapping at all. It's actually cubic: https://puu.sh/s541j/c2a71e1281.png
</details>
Hmm, with increased (height-)scale, it does look non-linear too :/
But it does resemble an S-curve that peaks upwards from `-1 .. 0` then levels through zero and dips downwards from `0 .. 1` - but not in a way that can be mirrored (unless your values are rounded oddly so it looks off). Closest match to a trend-line I can get is 6th order polynomial (up from at least 3rd order, which doesn't quite hit the center mark), but thats going to be a pain to implement... and probably not worth it either.
Suppose its best to keep the LUT for now, until someone comes up with a better idea.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418)*

**Comment by JosJuice** [comment]
From the graph, it looks a bit like you could use one linear mapping for -1 to 0.5 and another linear mapping for 0.5 to 1.

@JMC47 As far as I know, all Wii Guitar Hero games can use the slider bar except for Guitar Hero 3 and Guitar Hero Aerosmith (because they were released before Guitar Hero World Tour, and the Guitar Hero World Tour guitar was the first one to include a slider bar). Just to make sure before you start writing on the wiki, you might want to test if Guitar Hero Warriors of Rock supports it. The guitar that's made for Guitar Hero Warriors of Rock doesn't have a slider bar, but it would make sense if they still included slider bar support in the game for older guitars. (I think the band bundle for GH:WoR actually shipped with a GH5 guitar, which has a slider bar... Or am I wrong?)

I'm not really use what there is to note about this on the wiki, though. It's just a control on guitar controllers that some games use and some games don't use.

**Comment by Freyaday** [comment]
The way the slider bar is used in-game is that there are special notes that can be hit just by tapping the button for them when they cross the line instead of needing to strum (you can also strum). Tapping on the slider bar will hit these notes, as will sliding over to those notes when they cross the line, as well as tapping on the fret button. If there's none of these notes in Warriors of Rock, there's no way to test whether or not it works there, and if the included guitar doesn't have the slider bar, then I suspect none of them are in the charts for the songs.

---

Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418#-:-KVnvWfqjMP_gU-AYxM6:bsujqg7)*

**Comment by Freyaday** [comment]
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp, line 12 at r4](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418#-KVfAorfIOhqZhjc6VqK:-KVnwxVLJhTRaDJ6LYd3:bz7ftid) ([raw file](https://github.com/dolphin-emu/dolphin/blob/74115837a80fd088e6861683f40666fd39132a7b/Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp#L12)):*
<details><summary><i>Previously, BhaaLseN (BhaaL) wrote…</i></summary>
> Hmm, with increased (height-)scale, it does look non-linear too :/
> But it does resemble an S-curve that peaks upwards from `-1 .. 0` then levels through zero and dips downwards from `0 .. 1` - but not in a way that can be mirrored (unless your values are rounded oddly so it looks off). Closest match to a trend-line I can get is 6th order polynomial (up from at least 3rd order, which doesn't quite hit the center mark), but thats going to be a pain to implement... and probably not worth it either.
> Suppose its best to keep the LUT for now, until someone comes up with a better idea.
</details>
I redid the mappings for better accuracy and while they're strictly-increasing, they don't seem to follow any particular curve or formula: https://puu.sh/s7bul/e6801da2db.png A quintic function maps to them perfectly, but that's true of any six points (for n points, there's an n-1th order polynomial that exactly fits them).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418)*

**Comment by Freyaday** [comment]
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp, line 58 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418#-KVeUgFH2Qr7Mq2RCgzx:-KVnyQLkC15J-3mF71up:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/98f49403d84e1b63ad5fad110498b8728e33a8e8/Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp#L58)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> The variable should likely be changed to `m_sliderbar` or `m_slider_bar` as well
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418)*

**Comment by JosJuice** [comment]
Here are some slider notes in GH:WoR: https://www.youtube.com/watch?v=iZHnr_AnewM?t=2m22s (The video is probably not from the Wii version, but I have no reason to believe the charts would be different per console)

And you can test whether the slider bar works without having any slider notes, because it also has two other functions: When holding a sustain, it can be used in a way that's similar to the whammy bar, and in other cases, you can strum by tapping on it. (At least that's how I remember it working in GHWT.)

**Comment by BhaaLseN** [comment]
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp, line 12 at r4](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418#-KVfAorfIOhqZhjc6VqK:-KVoAumjS4RVcSF4dpUO:b-2sstsi) ([raw file](https://github.com/dolphin-emu/dolphin/blob/74115837a80fd088e6861683f40666fd39132a7b/Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp#L12)):*
<details><summary><i>Previously, Freyaday wrote…</i></summary>
> I redid the mappings for better accuracy and while they're strictly-increasing, they don't seem to follow any particular curve or formula: https://puu.sh/s7bul/e6801da2db.png A quintic function maps to them perfectly, but that's true of any six points (for n points, there's an n-1th order polynomial that exactly fits them).
</details>
This can't be that hard :P Maybe we're just overthinking it?
Anything but a linear or slightly-curved...curve makes no sense for an input device thats basically a slider; so I kinda doubt its a complex shape like that. Especially when there are 32 (or 31, if you dismiss the "not touching part") discrete values only.

If I could annoy you once more, would you be so kind to try out which value ranges still work for which frets? If `0x04` is _exactly_ the top fret, what happens if we input `0x03` or `0x05` etc? As for the hardware, I'd pretty much assume that going all the way to the top will just return a `-1`, then increase ever so slightly while sliding all the way down to `1`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418)*

**Comment by Freyaday** [comment]
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp, line 12 at r4](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418#-KVfAorfIOhqZhjc6VqK:-KVrMKOvCWW2vAIIZ17J:bk29nim) ([raw file](https://github.com/dolphin-emu/dolphin/blob/74115837a80fd088e6861683f40666fd39132a7b/Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp#L12)):*
<details><summary><i>Previously, BhaaLseN (BhaaL) wrote…</i></summary>
> This can't be that hard :P Maybe we're just overthinking it?
> Anything but a linear or slightly-curved...curve makes no sense for an input device thats basically a slider; so I kinda doubt its a complex shape like that. Especially when there are 32 (or 31, if you dismiss the "not touching part") discrete values only.
> 
> If I could annoy you once more, would you be so kind to try out which value ranges still work for which frets? If `0x04` is _exactly_ the top fret, what happens if we input `0x03` or `0x05` etc? As for the hardware, I'd pretty much assume that going all the way to the top will just return a `-1`, then increase ever so slightly while sliding all the way down to `1`.
</details>
http://wiibrew.org/wiki/Wiimote/Extension_Controllers/Guitar_Hero_(Wii)_Guitars
That's the table I'm using for frets. The top fret will only go to somewhere north of -1, but pressing the bottom fret returns exactly 1. Not pressing the slider bar returns 0. During my second round of testing for the more accurate table, the controller only returned one value per pad. I agree that the wacky value curve makes no sense, but that's what I'm getting. My guess is that it's some wacky hardware thing like some weird resistance or voltage curve (I have no idea how the slider bar works). I'll check out what the game can accept later, but it's certainly not ceil(Zr*16).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418)*

**Comment by BhaaLseN** [comment]
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp, line 12 at r4](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418#-KVfAorfIOhqZhjc6VqK:-KVtKU8X_yrCCzXB0ObH:bk3glq3) ([raw file](https://github.com/dolphin-emu/dolphin/blob/74115837a80fd088e6861683f40666fd39132a7b/Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp#L12)):*
The only thing that isn't implemented right now is
> You can hold two (touchbar) frets at once, as long as those two (touchbar) frets are adjacent, otherwise it will take the value of the lowest fret held (Eg, if you hold the highest and lowest fret, it reads 1F) Note that high/low means physically on the guitar neck (in musical terms I think it's the other way around, so I thought I'd better specify) 

...because you use `map.lower_bound`, which returns the lower _value_, not the lower _fret_ (which has the higher value).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4418)*

**Comment by JosJuice** [comment]
Sorry, I mis-clicked the assignment button.

@BhaaLseN We can't implement that feature unless we let the user map the controls in a way that makes it possible to provide two input values once. I don't see it's related to `lower_bound`.

**Comment by BhaaLseN** [comment]
Err, must've brainfarted somewhere...obviously the slider only returns one value at once, so nvm that comment :x

**Comment by JosJuice** [comment]
@Freyaday Can we get a rebase + squash of this PR?

I'd also like to know what you think about the idea I brought up about using one linear mapping for -1 to 0.5 and another linear mapping for 0.5 to 1.

**Comment by Freyaday** [comment]
I’ll take a look at how that mapping plays out. Since it’s only six values, I don’t think it’d be really any faster, and since I don’t have another guitar controller with a slider bar, I don’t know how much variance there is in these values (which is why it’s currently using a map as the LUT instead of an unordered_map; map allows for return values based off of ranges, whereas unordered_map, being a hashtable, is completely unaware of the actual values of anything). Rebasing will be a pain, due to how many commits I have to go through, but hopefully rerere will save me. It’ll take some time though.

**Comment by JosJuice** [comment]
One advantage of not using a six-value LUT is that it would let us properly map values that are in between frets. They don't matter for playing slider notes, but I'm guessing the game might take them into account when using the slider bar to change the sound of a sustain.

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

File: Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp [hunk 1]
```diff
@@ -3,6 +3,8 @@
 // Refer to the license.txt file included.
 
 #include "Core/HW/WiimoteEmu/Attachment/Guitar.h"
+#include <cstring>
+#include <map>
```
Comment (JosJuice) [inline Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp]: These two are supposed to be below `#include <cassert>`.

File: Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp [hunk 2]
```diff
@@ -11,11 +13,22 @@
...
+static const std::map<const ControlState, const u8> s_slider_bar_control_codes{
+    // values determined using a PS3 Guitar Hero 5 controller, which maps the touchbar to Zr on
+    // Windows
+    {0.0, 0x0F},        // not touching
+    {-0.4375, 0x04},    // top fret
+    {-0.097656, 0x0A},  // second fret
+    {0.203125, 0x12},   // third fret
+    {0.578125, 0x17},   // fourth fret
+    {1.0, 0x1F}         // bottom fret
+};
```
Comment (JosJuice) [inline Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp]: This is supposed to be below all the includes, inside `namespace WiimoteEmu`.

File: Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp [hunk 3]
```diff
@@ -83,8 +99,17 @@ void Guitar::GetState(u8* const data)
     gdata->sy = static_cast<u8>((y * 0x1F) + 0x20);
   }
 
-  // TODO: touch bar, probably not
-  gdata->tb = 0x0F;  // not touched
+  // slider bar
+  if (m_slider_bar->controls[0]->control_ref->BoundCount())
+  {  // if user has not mapped controls for slider bar, tell game it's untouched
```
Comment (JosJuice) [inline Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp]: This comment would probably be clearer if you put it above the line `gdata->sb = 0x0F;`

File: Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp (hunk 1 above)
Comment (Freyaday) [inline Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp]: done

File: Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp (hunk 2 above)
Comment (Freyaday) [inline Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp]: done

File: Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp (hunk 3 above)
Comment (Freyaday) [inline Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp]: done

Comment (JosJuice) [inline Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp]: And remove it from here, so that there isn't a redundant comment about it.

Comment (Freyaday) [inline Source/Core/Core/HW/WiimoteEmu/Attachment/Guitar.cpp]: also done

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

**Comment by JosJuice** [comment]
This PR looks fine to me, but I don't know enough about our controller code to give it a final LGTM.

**Comment by shuffle2** [comment]
something crazy happened here..

**Comment by JosJuice** [comment]
Oh, it did :( That's a shame... @Freyaday, do you still have the correct commit somewhere?

**Pull request closed without merging**